lib/grape/middleware/error.rb

```diff
@@ -119,21 +119,29 @@ def rescue_handler_for_any_class(klass)
         options[:all_rescue_handler] || :default_rescue_handler
       end
 
-      def run_rescue_handler(handler, error)
+      def run_rescue_handler(endpoint, handler, error)
         if handler.instance_of?(Symbol)
           raise NoMethodError, "undefined method '#{handler}'" unless respond_to?(handler)
 
           handler = public_method(handler)
         end
 
-        response = handler.arity.zero? ? instance_exec(&handler) : instance_exec(error, &handler)
+        response = (catch(:error) do
+          handler.arity.zero? ? endpoint.instance_exec(&handler) : endpoint.instance_exec(error, &handler)
+        end)
+
+        response = error!(response[:message], response[:status], response[:headers]) if error?(response)
```

Comment:
As now we are performing the handler under the endpoint instance, the methods called inside the `rescue_from` should be defined in the `inside_route.rb` file. I've experienced problems with this `error!` method:

The `error!` method. Is defined both times in `inside_routes.rb` and in the `error.rb`. However, the one defined in the `inside_routes.rb` is throwing an error and the one defined in the `error.rb` is just returning a rack response.

**Proposed Solution**: Catch the thrown error when performing the handler and if an error appears because someone calls the error! method, call the error! method from the `error.rb` again for returning the rack response.